File: 11-enumeration.rb
# Ruby Enumerable 將 map/select/sum/each 呢類資料處理串得好自然。
users = [
  { name: 'Ada', score: 95 },
  { name: 'Grace', score: 82 },
  { name: 'Linus', score: 70 }
]

names = users.map { |user| user[:name] }
passed = users.select { |user| user[:score] >= 80 }
total_score = users.sum { |user| user[:score] }

p names
p passed
puts total_score

puts '---'

# with_index(1) 可以令顯示編號由 1 開始。
users.each.with_index(1) do |user, index|
  puts "#{index}. #{user[:name]}"
end

puts '---'

score_by_name = users.to_h { |user| [user[:name], user[:score]] }

p score_by_name

File: 11-enumeration.py
# Python comprehension、generator expression 同 enumerate 係常見資料處理組合。
users = [
    {"name": "Ada", "score": 95},
    {"name": "Grace", "score": 82},
    {"name": "Linus", "score": 70},
]

names = [user["name"] for user in users]
passed = [user for user in users if user["score"] >= 80]
total_score = sum(user["score"] for user in users)

print(names)
print(passed)
print(total_score)

print("---")

# enumerate(..., start=1) 可以令顯示編號由 1 開始。
for index, user in enumerate(users, start=1):
    print(f"{index}. {user['name']}")

print("---")

score_by_name = {user["name"]: user["score"] for user in users}

print(score_by_name)
